Add route to link Incidents to Investigations

frenchy64 opened the pull request:
<!-- Specify linked issues and REMOVE THE UNUSED LINES -->

Close https://github.com/threatgrid/iroh/issues/2431

Extends `/incident/:id/link` to also support linking to Investigations.

<!--

Describe your PR for reviewers.
Don't forget to set correct labels (User Facing / Beta / Feature Flag)
If there is UI change please add a screen capture.
-->

<!-- UNCOMMENT THIS SECTION IF NEEDED
<a name="iroh-services-clients">[§](#iroh-services-clients)</a> IROH Services Clients
=====================================================================================

Put all informations that need to be communicated to IROH Services Clients.
Typically IROH-UI, ATS Integration, Orbital, etc...
 -->

<a name="qa">[§](#qa)</a> QA
============================

This PR broadens the scope of the `/incident/:id/link` route. For a body, it now accepts *one of* an `investigation_id` or `casebook_id` field, as well as some other optional keys (at the moment, just `tlp`).

Please follow these steps to create a relationship between an incident and an investigation, and verify it was created.

1. Visit `https://private.intel.int.iroh.site/index.html#/Incident/post_ctia_incident__id__link` (or equivalent)
2. Add a known incident id to the `id` param.
3. Add a body of the form `{ "investigation_id" : "<known-investigation-id>", ...}` with any relevant optional keys, then execute.
4. The response should be `200` and contain an `id` field. This is the relationship that was created.
5. Verify the `id` returned in step 4 corresponds to a relationship from the incident id to the investigation id.

The previous step should also still work for linking casebooks, via the `casebook_id` field. This behavior should be unchanged.

Please also verify providing zero or more than one of `investigation_id` or `casebook_id` fields in the body of the `/incident/:id/link` route triggers a 400 error response. Error messages should also be descriptive enough to diagnose the problem immediately.


<!-- UNCOMMENT THIS SECTION IF NEEDED
<a name="ops">[§](#ops)</a> Ops
===============================

  Specify Ops related issues and documentation
- Config change needed: threatgrid/tenzin#
- Migration needed: threatgrid/tenzin#
- How to enable/disable that feature: (ex remove service from `bootstrap.cfg`, add scope to org)
-->

<!-- UNCOMMENT THIS SECTION IF NEEDED
<a name="documentation">[§](#documentation)</a> Documentation
=============================================================

  Public Facing documentation section;
- Public documentation updated needed: threatgrid/iroh-ui#
  See internal [doc file](./services/iroh-auth/doc/public-doc.org)
 -->

<a name="release-notes">[§](#release-notes)</a> Release Notes
=============================================================

<!-- REMOVE UNUSED LINES -->

```
intern: Add route to link Incidents to Investigations
```

<a name="squashed-commits">[§](#squashed-commits)</a> Squashed Commits
======================================================================

vemv commented on test/ctia/entity/relationship_test.clj:
```diff
@@ -172,11 +178,84 @@
+                  :headers {"Authorization" "45c1f5e3f05d0"})
+             _ (is (= 200 relationship-status))]
+
+         (is (= (:id investigation-body)
```
There's the risk of a nil vs nil comparison here

vemv commented on test/ctia/entity/relationship_test.clj:
```diff
@@ -172,11 +178,84 @@
+                                         "foouser"
+                                         "foogroup"
+                                         "user")
+     (testing "Incident <-> Investigation link"
```
The `testing` block is quite large, but in this case it seems justified since it's a stateful test that "tells a story" out of multiple requests.

Maybe expand `"Incident <-> Investigation link"` to summarize that story, else one finds himself reading a large `let` without knowing exactly what to expect.

msprunck commented:
I think there is a misunderstanding with the current route:

```
POST /incident/:incident_id/link
{:casebook_id "casebook_id"}
```

The path param defines the resource ID on which the `link` action is performed, in this case the incident ID.

I would just add a `investigation_id` in the body of the current route. IMO There is no need to add a new route.

frenchy64 commented:
Yes that makes sense @msprunck. I'll do that!

frenchy64 commented:
I believe I've addressed all feedback. Please take another look.

msprunck commented on src/ctia/entity/relationship.clj:
```diff
@@ -71,51 +72,75 @@
+                                                      [:casebook_id
+                                                       :investigation_id])]
+                               (when-not (= 1 (count types))
+                                 (bad-request! {:error "Missing source type"}))
```
```suggestion
                                 (bad-request! {:error "Please provide a casebook_id OR an investigation_id"}))
```

The current error message is not very explicit.

msprunck replied:
Could you test the cases where:
1. `casebook_id` and `investigation_id` are provided
2. none of them are provided

frenchy64 commented:
Added some tests, and refactored a bit to make them easier to generate in the future.

```clojure
body = {}
; 400 response with message:
;; Please provide exactly one of the following fields: casebook_id, investigation_id
;; None provided.

body = {:casebook_id ":", :investigation_id ":"}
; 400 response with message:
;; Please provide exactly one of the following fields: casebook_id, investigation_id
;; Provided: casebook_id, investigation_id
```

frenchy64 commented:
Also added QA instructions. Unfortunately I pushed this branch directly to the threatgrid repo, so I triggered a double build. The build is also very flaky today. Will merge in the next few hours.

vemv commented on test/ctia/entity/relationship_test.clj:
```diff
@@ -195,70 +194,88 @@
+                    ::no-target-ref))
+             "The new Relationship targets the incident")
+         (is (= (or relationship-response
+                    ::no-relationship-response)
```
Nice trick!

hshahman commented:
> Please also verify providing zero or more than one of investigation_id or casebook_id fields in the body of the /incident/:id/link route triggers a 400 error response. Error messages should also be descriptive enough to diagnose the problem immediately.

There are 2 issues wrt above statement.
1. Sending a request with 2 investigation_id to `/link` endpoint is going through and response is 201, which is not expected. Here is the body:
```
{
  "investigation_id": "https://private.intel.test.iroh.site:443/ctia/investigation/investigation-d8c17fe0-d78b-4ac1-b08e-8bfbd549861f",
  "investigation_id": "https://private.intel.test.iroh.site:443/ctia/investigation/investigation-17873611-8caa-4053-ac65-f1d480138d2a",
  "tlp": "green"
}
```


Request:
```
curl -X POST "https://private.intel.test.iroh.site/ctia/incident/incident-1df8534a-1e3d-4802-ab13-b09a3395556e/link" -H "accept: application/json" -H "authorization: Bearer <JWT_TOKEN>" -H "Content-Type: application/json" -d "{ \"investigation_id\": \"https://private.intel.test.iroh.site:443/ctia/investigation/investigation-d8c17fe0-d78b-4ac1-b08e-8bfbd549861f\", \"investigation_id\": \"https://private.intel.test.iroh.site:443/ctia/investigation/investigation-17873611-8caa-4053-ac65-f1d480138d2a\", \"tlp\": \"green\"}"
```

Response contains only one relationship between the incident_id and the **_second_** investigation_id:
```
{
  "schema_version": "1.0.16",
  "target_ref": "https://private.intel.test.iroh.site:443/ctia/incident/incident-1df8534a-1e3d-4802-ab13-b09a3395556e",
  "type": "relationship",
  "source_ref": "https://private.intel.test.iroh.site:443/ctia/investigation/investigation-17873611-8caa-4053-ac65-f1d480138d2a",
  "id": "https://private.intel.test.iroh.site:443/ctia/relationship/relationship-64fc1e84-984b-4d3f-95d5-e29ab5247ee0",
  "tlp": "green",
  "groups": [
    "cc6a35bc-1739-4fcd-a285-aa95adbd5e41"
  ],
  "timestamp": "2020-07-31T18:25:18.239Z",
  "owner": "idb-amp:f6b8d1c4-189b-43eb-b1ba-191d53357a8e",
  "relationship_type": "related-to"
}
```

2. Sending a query to the `/link` endpoint with no investigation_id in the body is returning a 401 as expected, but there is no error message. I think a descriptive error message is required here as mentioned in the QA instructions by Ambrose.

frenchy64 commented:
Wow, I did not think of that case! I had no idea JSON allowed duplicate keys in maps.

frenchy64 commented:
Thanks @hshahman for such a thorough review. After some discussion, we decided that duplicate keys in JSON maps are out of scope for our API (or at least, future work), so we'll just fix 2 at the moment. I'll refine the QA instructions on the next PR.

hshahman commented:
hmmm, ok...but may I ask  why it is out of scope of our APIs? It doesn't make sense to me why duplicate keys are allowed.

frenchy64 commented:
My interpretation is that you'll find this problem with all SecureX/CTR APIs. This is because none of our code deals with raw JSON, instead we use the Clojure representation, which clobbers duplicate keys.

To tackle the issue, we'd probably need our own JSON parser that checks for duplicate keys. You could imagine that response times would be slower doing these extra checks and the maintenance burden is higher.

I think GB's opinion is that all that work is not worth the benefits of fixing the problem. But that's just my interpretation, feel free to ask him if you want more info.

hshahman commented:
@frenchy64 thank you for the details above. I'm going to file a new ticket for this issue:
> 2. Sending a query to the /link endpoint with no investigation_id in the body is returning a 401 as expected, but there is no error message. I think a descriptive error message is required here as mentioned in the QA instructions by Ambrose.

So we can pass this ticket since it's almost working as expected.

hshahman commented:
Verified in rel-v1.54
